File: p0100/p0045_triangular_pentagonal_hexagonal/triangular_pentagonal_hexagonal.cpp
#include <iostream>
#include <cmath>

// Check if a number is pentagonal
bool isPentagonal(long long number) {
    long long s = static_cast<long long>(std::round(std::sqrt(1 + 8 * number)));
    return s * s == (1 + 8 * number) && (s - 1) % 2 == 0;
}

// Check if a number is triangular
bool isTriangular(long long number) {
    long long s = static_cast<long long>(std::round(std::sqrt(1 + 24 * number)));
    return s * s == (1 + 24 * number) && (1 + s) % 6 == 0;
}

// Brute force search for next number that is triangular, pentagonal, and hexagonal
long long tphBruteForce() {
    for (long long k = 144;; ++k) {
        long long hk = k * (2 * k - 1);  // hexagonal number formula
        if (isPentagonal(hk) && isTriangular(hk)) {
            return hk;
        }
    }
}

int main() {
    long long result = tphBruteForce();
    std::cout << "Hexagonal number: " << result << std::endl;
    return 0;
}

File: p0100/p0045_triangular_pentagonal_hexagonal/triangular_pentagonal_hexagonal.py
import itertools
import math


def isPentagonal(number):
    s = int(round(math.sqrt(1 + 8 * number)))
    return s**2 == (1 + 8 * number) and (s - 1) % 2 == 0


def isTriangular(number):
    s = int(round(math.sqrt(1 + 24 * number)))
    return s**2 == (1 + 24 * number) and (1 + s) % 6 == 0


def tphBruteForce():
    for k in itertools.count(144, 1):
        hk = 2 * (2 * k - 1)
        if isPentagonal(hk) and isTriangular(hk):
            return k


if __name__ == '__main__':
    k = tphBruteForce()
    print(k)
    print(k * (2 * k - 1))
